ONSdigital/eq-questionnaire-launcher
# Authenticate SDS requests

### What is the context of this PR?
This PR updates launcher to authenticate SDS requests with GCP. It follows a similar approach to the previously merged [runner pr](https://github.com/ONSdigital/eq-questionnaire-runner/pull/1172)

The oauth2 library wasn't working with go version 1.16 so this PR has also updated to the latest stable version of go and updated packages.

### How to review
**Without GCP authentication**
Simply check that switching to this PR, without making any changes, still allows you to launch supplementary data surveys locally while the mock SDS is running

**GCP locally**
Similar to runner

1. If you already have a service account json file from testing the runner PR, you can use this - otherwise
   - Log into the GCP console in Chrome
   - Create or use an existing eq-sandbox project, add a service account and client id under APIs and Services > Credentials.
   - Download the service account json file and place it somewhere in the project.
2. Set the following environment variables for launcher
   - `SDS_OAUTH2_CLIENT_ID=<the Client Id in the GCP project>`
   - `GOOGLE_APPLICATION_CREDENTIALS=<path to the service account json>`
   - `OIDC_TOKEN_BACKEND=gcp`
   - If you're using a JetBrains tool, you can set these in your run configuration
   - If you're using VSCode, go to `.vscode/settings.json` and add the following
    ```
    {
        "go.testEnvVars": {
             "SDS_OAUTH2_CLIENT_ID": "<the Client Id in the GCP project>",
             "GOOGLE_APPLICATION_CREDENTIALS": "<path to the service account json>"
             "OIDC_TOKEN_BACKEND"="gcp"
         },
    }
    ```
    alternatively for `OIDC_TOKEN_BACKEND` you could change the default in `settings.go`
3. Open Launcher and pick `test_supplementary_data`
4. You should notice `Succesfully generated GCP ID token` in the logs
5. Refresh the page and pick the same schema, this time you should see `Found cached GCP ID token source`
6. To test the TTL Cache, you can change the OIDC_TOKEN_VALIDITY_IN_SECONDS and leeway in settings.go, or pass them in as environment variables, and pick a shorter time frame. Note that the cache is only cleared once per minute so if you want to test a shorter amount of time then change `time.Minute` in `oidc.go` to `time.Second` or similar

**GCP pipeline**
1. Follow the steps from the [PR supporting the mock SDS ](https://github.com/ONSdigital/eq-pipelines/pull/503) to create your own staging environment.
2. runner.image_tag can just be latest as the runner PR is merged
3. If you test prior to changing anything, following the steps on the PR should show launcher not working, but runner supporting supplementary data via quicklaunch
3. Then set the launcher image tag to `authenticate-with-sds` and update the pipeline
4. Launcher and runner should both now work
5. Note that you don't need to do anything with oidc_token_backend in the pipelines as this is set in the terraform repo and defaults to gcp
